### DISCUSSION TITLE
Java springboot packaging error

### DISCUSSION THREAD

**Discussion by Bestbbb** [google/or-tools#3553] [Build and configuration issues] [answered]
OR-TOOLS runs in the development environment normally,but when I package or install it  in the springboot jar,it will show the error:
_java.nio.file.NoSuchFileException: /BOOT-INF/lib/ortools-win32-x86-64-9.4.1874.jar!/ortools-win32-x86-64_

I try all the methods metioned before,but it cannot work

**Comment by henszey** [comment, marked as answer]
This is because Spring Boot does not shade the included jars the same way other builds do, see:

https://docs.spring.io/spring-boot/docs/current/reference/html/executable-jar.html#appendix.executable-jar.nested-jars

There is an issue that describes this with several solutions.

Attached what I have used which is different what the issues show as solutions


[SpringBootOrToolsNativeLoader.zip](https://github.com/google/or-tools/files/10079481/SpringBootOrToolsNativeLoader.zip)

**Reply by Bestbbb** [reply]
I cannot find the class CustomOrToolsNativeLoader, is it existed in other maven?And how can I use this solution,shall I just put  in the current project or repackage in the or-tools jar?

**Reply by Bestbbb** [reply]
I try a solution,I use the maven-shade plugin instead of spring-boot-maven-plugin,and finally it works.

**Reply by nishant-fareye** [reply]
> I cannot find the class CustomOrToolsNativeLoader, is it existed in other maven?And how can I use this solution,shall I just put in the current project or repackage in the or-tools jar?

You just need to replace **CustomOrToolsNativeLoader** with **SpringBootOrToolsNativeLoader**.
I guess it's just a typo in the provided solution.

**Reply by henszey** [reply]
Reuploaded with the typo fixed incase anyone else stumbles upon this thread.

**Reply by nishant-fareye** [reply]
@Bestbbb can you please share if the steps mentioned over here are correct to switch from spring-boot-maven-plugin to maven-shade plugin?
https://stackoverflow.com/questions/33687492/spring-boot-with-maven-shade-plugin-controllers-not-mapped-404-error#:~:text=migrate%20from%20spring%2Dboot%2Dmaven%20plugin%20to%20maven%2Dshade%20plugin

If not can you please share the same

**Reply by Bestbbb** [reply]
I shared it,it is a little different.

**Reply by nishant-fareye** [reply]
> I try a solution,I use the maven-shade plugin instead of spring-boot-maven-plugin,and finally it works.

for me as well the solution provided in **SpringBootOrToolsNativeLoader**  only with **maven-shaded-plugin** and not with **spring-boot-maven-plugin**
@henszey just checking if the solution also works with **spring-boot-maven-plugin**?

**Reply by Sunzt1995** [reply]
I use this method, but this error occurred：
java.nio.file.ProviderNotFoundException: Provider not found
at java.nio.file.FileSystems.newFileSystem(FileSystems.java:407)
at com.xiaoju.automarket.pds.strategy.domain.loaddispatch.utils.SpringBootOrToolsNativeLoader.visitUriRecursionHelper(SpringBootOrToolsNativeLoader.java:87)

**Comment by ClaudioConsolmagno** [comment]
The best solution for me was described in this ticket by @C-Otto  https://github.com/magneticflux-/ortools-java/issues/5#issuecomment-1075407198 . Just add this to your build file:

```
bootJar {
    requiresUnpack '**/ortools-darwin*.jar', '**/ortools-win32*.jar', '**/ortools-linux*.jar'
}
```

**Reply by rmustard** [reply]
I used this solution in maven using the spring-boot-maven-plugin
```xml
<build>
  <plugins>
    <plugin>
      <groupId>org.springframework.boot</groupId>
      <artifactId>spring-boot-maven-plugin</artifactId>
      <version>${project.parent.version}</version>
      <configuration>
        <requiresUnpack>
            <dependency>
              <groupId>com.google.ortools</groupId>
              <artifactId>ortools-java</artifactId>
            </dependency>
            <dependency>
              <groupId>com.google.ortools</groupId>
              <artifactId>ortools-linux-x86-64</artifactId>
            </dependency>
            <dependency>
              <groupId>com.google.ortools</groupId>
              <artifactId>ortools-darwin-x86-64</artifactId>
            </dependency>
            <dependency>
              <groupId>com.google.ortools</groupId>
              <artifactId>ortools-win32-x86-64</artifactId>
            </dependency>
            <dependency>
              <groupId>com.google.ortools</groupId>
              <artifactId>ortools-linux-aarch64</artifactId>
            </dependency>
            <dependency>
              <groupId>com.google.ortools</groupId>
              <artifactId>ortools-darwin-aarch64</artifactId>
            </dependency>
        </requiresUnpack>
      </configuration>
    </plugin
  </plugins>
</build>
```

**Reply by ENCHIGO** [reply]
thanks ! I fixed it by your way

**Reply by Mizux** [reply]
ref: https://docs.spring.io/spring-boot/docs/current/maven-plugin/reference/htmlsingle/#packaging.repackage-goal.parameter-details.requires-unpack
